src/game_events/handlers.cpp

```diff
@@ -236,6 +236,28 @@ struct filter_unit : public event_filter {
 	bool first_;
 };
 
+namespace {
+/**
+ * The event data passed to filter_attack has the details of the weapons used, however if the
+ * defender has no weapon of the correct range then a placeholder is included. This undoes that,
+ * returning nullptr instead of turning the placeholder into an attack_type.
+ *
+ * This also handles choosing the first or second weapon.
+ *
+ * Though it may seem wasteful to use std::shared_ptr instead of std::optional, it's necessary.
+ * matches_filter() could potentially call a WFL formula, which would call shared_from_this().
+ */
+std::shared_ptr<const attack_type> get_attack_event_weapon(const config& event_data, bool first) {
+	// The placeholders are constructed in attack.cpp's attack::fire_event, and contain only the
+	// name="none", so can be recognised by checking for a range attribute.
+	auto attack_cfg = event_data.optional_child(first ? "first" : "second");
```

Comment:
best to let placeholder and use that https://github.com/wesnoth/wesnoth/pull/7539 for resolve problem of special_id_active matches in placeholder weapon